### PR TITLE
536 update fields for adding project

### PR DESCRIPTION
**Pull request by momimomo**
#536, #563 and first bug from #586 ready.

### CODE REVIEW

**Comment by rkrupinski** [comment]
@momimomo run `yarn run bundle` from `<projectDir>/polymer`:

```
src/pages/app/cluster-reporting/response-parameters/partners/projects.html_script_0.js
  line 45  col 49  Missing semicolon.
  line 45  col 39  requireSemicolons: Missing semicolon after statement

  ⚠  2 warnings


src/elements/cluster-reporting/add-response-plan-modal.html_script_0.js
  line 222  col 13  'self' is not defined.

  ⚠  1 warning


src/elements/cluster-reporting/planned-action/projects/creation-modal.html_script_0.js
  line 111  col 58  Strings must use singlequote.
  line 111  col 20  validateQuoteMarks: Invalid quote mark found
  line 142  col 22  Duplicate key 'detailsOpened'.
  line 161  col 64  requireSemicolons: Missing semicolon after statement
  line 161  col 66  Missing semicolon.
  line 162  col 70  Missing semicolon.
  line 162  col 68  requireSemicolons: Missing semicolon after statement
  line 166  col 60  requireSemicolons: Missing semicolon after statement
  line 166  col 62  Missing semicolon.
  line 196  col 80  Expected an assignment or function call and instead saw an expression.
  line 196  col 84  Missing semicolon.
  line 196  col 81  requireSemicolons: Missing semicolon after statement
  line 290  col 14  Missing semicolon.
  line 290  col 12  requireSemicolons: Missing semicolon after statement
  line 332  col 27  Missing semicolon.
  line 332  col 23  requireSemicolons: Missing semicolon after statement
  line 354  col 51  requireSemicolons: Missing semicolon after statement
  line 354  col 62  Missing semicolon.

  ⚠  18 warnings


src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html
  line 90  col 34  The id value [ [[field.id]] ] must be unique.
  line 93  col 40  The id value [ [[field.id]] ] must be unique.

  ⚠  2 warnings


src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html_script_0.js
  line 20  col 10  disallowMultipleLineBreaks: , and _add should have at most 2 line(s) between them
  line 43  col 31  Expected '===' and instead saw '=='.
  line 52  col 31  Expected '===' and instead saw '=='.
  line 60  col 25  Missing semicolon.
  line 60  col 22  requireSemicolons: Missing semicolon after statement
  line 66  col 28  Missing semicolon.
  line 66  col 29  Unnecessary semicolon.
  line 66  col 24  requireSemicolons: Missing semicolon after statement
  line 66  col 27  requireSpaceAfterKeywords: One space required after "return" keyword
  line 67  col 34  requireSpaceBetweenArguments: Missing space between , and [
  line 71  col 9   Missing semicolon.
  line 71  col 7   requireSemicolons: Missing semicolon after statement

  ⚠  12 warnings

[16:06:59] 'output' errored after 6.96 s
[16:06:59] Error in plugin 'gulp-htmlhint'
```

**Comment by rkrupinski** [comment]
Also, @momimomo , 50% of the PR is whitespace. Could you please either disable `editor.formatOnSave` in VSCode or configure it according to the convention? :)

**Comment by momimomo** [comment]
`editor.formatOnSave` is disabled, so all whitespace changes are intentional manual formatting… please correct me if it was unnecessary but I wanted to keep the same code style as in other components.

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 1]
```diff
@@ -101,170 +213,438 @@
...
             </div>
+            <iron-collapse
+              id="collapse"
+              opened="{{detailsOpened}}">
+              <div class="app-grid">
+                <etools-single-selection-menu
+                  id="agencyAcronyms"
+                  class="item validate"
+                  label="Name of Agency (acronyms)"
+                  options="[[agencyAcronyms]]"
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: `agencyAcronyms` is not defined

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 2]
```diff
@@ -101,170 +213,438 @@
...
+                  selected="{{selectedAcronym}}"
+                  on-iron-activate="_validate"
+                  trigger-value-change-event
+                  always-float-label>
+                </etools-single-selection-menu>
+
+                <etools-single-selection-menu
+                  class="item validate"
+                  label="Type of Agency"
+                  options="[[agencyTypes]]"
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: `agencyTypes` is not defined

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 3]
```diff
@@ -101,170 +213,438 @@
...
+                      <paper-radio-button name="fa">FA</paper-radio-button>
+                      <paper-radio-button name="no">No</paper-radio-button>
+                    </paper-radio-group>
+                  </labelled-item>
+                </div>
+
+                <etools-single-selection-menu
+                  class="item validate"
+                  label="Prioritization Classification"
+                  options="[[classifications]]"
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: `classifications` is not defined

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 4]
```diff
@@ -101,170 +213,438 @@
...
+                  selected="{{selectedClassification}}"
+                  on-iron-activate="_validate"
+                  trigger-value-change-event
+                  always-float-label>
+                </etools-single-selection-menu>
+
+                <etools-single-selection-menu
+                  class="item validate"
+                  label="Project code in HRP"
+                  options="[[projectCodes]]"
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: `projectCodes` is not defined

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 5]
```diff
@@ -418,8 +851,81 @@ <h2>[[_computeTitle(edit)]]</h2>
...
+        return projects.map(function(project) {
+          return {id: project.id, title: project.name};
+        });
+      },
+
+      // _onOpenedChanged: function (opened) {
+      //   if (opened) {
+      //     this.set('mode', '');
+      //   }
+      // },
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: Needed?

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 6]
```diff
@@ -434,6 +940,14 @@ <h2>[[_computeTitle(edit)]]</h2>
         e.target.validate();
       },
 
+      _onValueChanged: function (e) {
+        this._updateValue(e.target);
+      },
+
+      _updateValue: function (el) {
+        this.set(el.dataset.path, el.selected);
+      },
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: `_onValueChanged ` and `_updateValue` can be replaced by simply binding to `selected="{{data.part}}"`

File: polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html [hunk 7]
```diff
@@ -452,23 +966,33 @@ <h2>[[_computeTitle(edit)]]</h2>
...
+        this.set('data', {});
+        this.set('projectDetails', {});
+        this.set('errors', {});
+      },
 
-        var valid = [
-          this._fieldsAreValid(),
-          this._dateRangeValid('.start-date', '.end-date'),
-        ].every(Boolean);
+      _adjustPosition: function (e) {
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/creation-modal.html]: Redundant (see `ModalBehavior`)

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 8]
```diff
@@ -0,0 +1,195 @@
+<link rel="import" href="../../../../../bower_components/polymer/polymer.html">
+<link rel="import" href="../../../../../bower_components/iron-flex-layout/iron-flex-layout-classes.html">
+<link rel="import" href="../../../../../bower_components/app-layout/app-grid/app-grid-style.html">
+<link rel="import" href="../../../../../bower_components/paper-button/paper-button.html">
+<link rel="import" href="../../../../../bower_components/iron-icons/iron-icons.html">
+<link rel="import" href="../../../../../bower_components/paper-icon-button/paper-icon-button.html">
+<link rel="import" href="../../../../../bower_components/etools-searchable-multiselection-menu/etools-single-selection-menu.html">
+<link rel="import" href="../../../../../bower_components/paper-tooltip/paper-tooltip.html">
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: Unused?

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+<link rel="import" href="../../../../polyfills/es6-shim.html">
+<link rel="import" href="../../../../behaviors/utils.html">
+<link rel="import" href="../../../../redux/store.html">
+<link rel="import" href="../../../../styles/buttons.html">
+<link rel="import" href="../../../../styles/shared-styles.html">
+<link rel="import" href="../../../json-field.html">
+<link rel="import" href="../../../etools-prp-ajax.html">
+<link rel="import" href="../../../etools-prp-reset.html">
+<link rel="import" href="../../../etools-prp-permissions.html">
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: Unused?

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+      labelled-item {
+        padding: 8px 0;
+      }
+
+    </style>
+
+    <etools-prp-permissions
+      permissions="{{permissions}}">
+    </etools-prp-permissions>
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: Unused?

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+
+    <template
+      is="dom-repeat"
+      items="[[customFields]]"
+      as="field"
+      restamp="true">
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: `dom-repeat` does not make use of `restamp`

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+<link rel="import" href="../../../../../bower_components/paper-button/paper-button.html">
+<link rel="import" href="../../../../../bower_components/iron-icons/iron-icons.html">
+<link rel="import" href="../../../../../bower_components/paper-icon-button/paper-icon-button.html">
+<link rel="import" href="../../../../../bower_components/etools-searchable-multiselection-menu/etools-single-selection-menu.html">
+<link rel="import" href="../../../../../bower_components/paper-tooltip/paper-tooltip.html">
+<link rel="import" href="../../../../../bower_components/paper-input/paper-input.html">
+
+<link rel="import" href="../../../../polyfills/es6-shim.html">
+<link rel="import" href="../../../../behaviors/utils.html">
+<link rel="import" href="../../../../redux/store.html">
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: Unused?

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        Add Custom Field
+    </paper-button>
+  </template>
+    <script>
+      Polymer({
+        is: 'custom-fields-widget',
+
+        behaviors: [
+          App.Behaviors.UtilsBehavior,
+          App.Behaviors.ReduxBehavior,
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: Unused?

File: polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        ],
+
+        properties: {
+          fieldId: Number,
+          customFields: {
+            format: Array,
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/planned-action/projects/custom-fields-widget.html]: `format` -> `type`

File: polymer/src/elements/cluster-reporting/project-details.html [hunk 15]
```diff
@@ -86,7 +106,126 @@
...
             </li>
 
+
+
           </ul>
+          <div class="app-grid">
+          <template
+          is="dom-repeat"
+          items="[[projectData.custom_fields]]"
+          restamp="true"
```
Comment (rkrupinski) [inline polymer/src/elements/cluster-reporting/project-details.html]: `restamp` is obsolete

**Comment by momimomo** [comment]
I noticed that I still left some bugs with custom fields - adding fixes to pull request.

**Comment by momimomo** [comment]
editing custom project removes all previous project. Will have to change backend endpoint and request. I'll figure it out with @maciej-jaworski .

**Comment by momimomo** [comment]
with patch method I'm also sending id in object, so i looks like:


and Im sending it under:


That deletes previous entries.

**Comment by momimomo** [comment]
Endpoint changes and edit mode fixes are in `536-assigning-partner-to-ocha-project`

**Pull request merged**